Separate vesting from auth, add custom vesting schedules (#5040)
cosmos/cosmos-sdk: x/auth/spec/05_vesting.md

## Intro and Requirements

> This specification defines the vesting account implementation that is used by the Cosmos Hub. The requirements for this vesting account is that it should be initialized during genesis with a starting balance `X` and a vesting end time `T`. A vesting account may be initialized with a vesting start time `T'` and a number of vesting periods `P`. If a vesting start time is included, the vesting period will not begin until start time is reached. If vesting periods are included, the vesting will occur over the specified number of periods.

Reviewer: Markdown should be wrapped at 80-120 chars. @marbar3778 is there a linter for this?

Reviewer 2: yeah, worst case you can define break lines on the IDE's settings

> This specification defines the vesting account implementation that is used by the Cosmos Hub. The requirements for this vesting account is that it should be initialized during genesis with a starting balance `X` and a vesting end time `T`. A vesting account may be initialized with a vesting start time `T'` and a number of vesting periods `P`. If a vesting start time is included, the vesting period will not begin until start time is reached. If vesting periods are included, the vesting will occur over the specified number of periods.

Reviewer: ```suggestion
This specification defines the vesting account implementation that is used by the Cosmos Hub. The requirements for vesting accounts is that they should be initialized during genesis with a starting balance `X` and a vesting end time `T`. A vesting account may be initialized with a vesting start time `T'` or a number of vesting periods `P`. If a vesting start time is included, the vesting period will not begin until start time is reached. If vesting periods are included, the vesting will occur over the specified number of periods.
```

Reviewer: Also, I realized we use `T` for multiple values, let's change them to `ST` for start time and `ET` for end time?

> For all vesting accounts, the owner of the vesting account is able to delegate and undelegate from validators, however they cannot transfer coins to another account until those coins are vested. This specification allows for three different kinds of vesting:

Reviewer: ditto on wrapping

type DelayedVestingAccount struct {
    BaseVestingAccount
}

// VestingPeriod defines a length of time and amount of coins that will vest
type VestingPeriod struct {

>   PeriodLength int64 // length of the period, in seconds

Reviewer 2: ```suggestion
  Length int64 // length of the period, in seconds
```

>   VestingAmount Coins // amount of coins vesting during this period

Reviewer 2: ```suggestion
  Amount Coins // amount of coins vesting during this period
```

}
```

### Periodic Vesting Accounts
Periodic vesting accounts require calculating the coins released during each period for a given block time `T`. Note that multiple periods could have passed when calculating GetVestedCoins, so we must iterate over each period until the end of that period is after `T`.

Set `CT := StartTime`
Set `V' := 0`
For each Period P:
  1. Compute `X := T - CT`
  2. IF `X >= P.Length`
    a. Compute `V' += P.Amount`
    b. Compute `CT += P.Length`
    ELSE break
  3. Compute `V := OV = V'`

```go
func (pva PeriodicVestingAccount) GetVestedCoins(t Time) Coins {
  if t < pva.StartTime {
    return ZeroCoins
  }
  ct := pva.StartTime // The time of the current period start
  vested := 0
  periods = pva.GetPeriods()
  for i := 0; i < len(periods); i++ {
    x := t - ct

>     if x>= periods[i].Length {

Reviewer 2: you can remove the `else` with the following:
```go
if t - ct < periods[i].Length {
    break
}
```

}
```

### Periodic Vesting Accounts

> Periodic vesting accounts require calculating the coins released during each period for a given block time `T`. Note that multiple periods could have passed when calculating GetVestedCoins, so we must iterate over each period until the end of that period is after `T`.

Reviewer: ```suggestion
Periodic vesting accounts require calculating the coins released during each period for a given block time `T`. Note that multiple periods could have passed when calling `GetVestedCoins`, so we must iterate over each period until the end of that period is after `T`.
```

Set `CT := StartTime`

Set `V' := 0`

For each Period P:

  1. Compute `X := T - CT`
  2. IF `X >= P.Length`
      1. Compute `V' += P.Amount`
      2. Compute `CT += P.Length`
      3. ELSE break

>   3. Compute `V := OV = V'`

Reviewer: This looks malformed and incorrect?

```go
func (pva PeriodicVestingAccount) GetVestedCoins(t Time) Coins {
  if t < pva.StartTime {
    return ZeroCoins
  }

>   ct := pva.StartTime // The time of the current period start

Reviewer: isn't `CT` the start of the vesting schedule?

    Notice how we have an excess amount of `DV`.

### Periodic Vesting

A vesting account is created where 100 tokens will be released over 1 year, with 1/4 of tokens vesting each quarter. The vesting schedule would be as follows:

> ```json

Reviewer: Can we avoid JSON? Just use simple text/YAML.

{
  "vesting_periods": [
    {
      "period_length": 7884000,
      "vesting_amount": [
        {
        "denom": "stake",
        "amount": "25"
      }
    ]
    },
    {
      "period_length": 7884000,
      "vesting_amount": [
        {
        "denom": "stake",
        "amount": "25"
      }
    ]
    },
    {
      "period_length": 7884000,
      "vesting_amount": [
        {
        "denom": "stake",
        "amount": "25"
      }
    ]
    },
    {
      "period_length": 7884000,
      "vesting_amount": [
        {
        "denom": "stake",
        "amount": "25"
      }
    ]
    }
  ]
}
```

```
OV = 100
DF = 0
DV = 0
BC = 100
V = 100
V' = 0
```

1. Immediately receives 1 coin

    ```
    BC = 101
    ```

2. Vesting period 1 passes, 25 coins vest

    ```
    V = 75
    V' = 25
    ```

> 3. During vesting period 2, 5 coins are transfered and 5 coins are delegated

Reviewer: It's been a while since I wrote this, but having trouble following along here, so correct me if I'm wrong.

At this point, assuming we're in period 2 and it has not passed, then:

```
X ← min(max(75 - 0, 0), 5)
  ← min(max(75, 0), 5)
  ← min(75, 5)
  ← 5

DV ← DV + 5
   ← 5
```
